addons/hr_holidays/models/hr_leave_allocation.py

```diff
@@ -84,18 +93,95 @@ def _default_holiday_status_id(self):
+
+        today = to_tz(fields.Datetime.from_string(fields.Datetime.now()), self.env.user.tz).date()
+
+        holidays = self.search([('accrual', '=', True), ('state', '=', 'validate'),
+                                '|', ('date_to', '=', False), ('date_to', '>', fields.Datetime.now()),
+                                '|', ('nextcall', '=', False), ('nextcall', '<=', today)])
+
+        for holiday in holidays:
+            values = {}
+
+            leaves = holiday.number_per_interval
+            if holiday.unit_per_interval == 'hours':
+                leaves = holiday.number_per_interval / holiday.employee_id.resource_calendar_id.hours_per_day
+
+            delta = relativedelta(days=0)
+
+            if holiday.interval_unit == 'weeks':
+                delta = relativedelta(weeks=holiday.interval_number)
+            if holiday.interval_unit == 'months':
+                delta = relativedelta(months=holiday.interval_number)
+            if holiday.interval_unit == 'years':
+                delta = relativedelta(years=holiday.interval_number)
+
+            values['nextcall'] = (holiday.nextcall if holiday.nextcall else today) + delta
+
+            period_start = datetime.combine(today, time(0, 0, 0)) - delta
+            period_end = datetime.combine(today, time(0, 0, 0))
+
+            # We have to check when the employee has been created
+            # in order to not allocate him/her too much leaves
+            creation_date = fields.Datetime.from_string(holiday.employee_id.create_date)
+
+            # If employee is created after the period, we cancel the computation
+            if period_end <= creation_date:
+                holiday.write(values)
+                continue
+
+            # If employee created during the period, taking the date at which he has been created
+            if period_start <= creation_date:
+                period_start = creation_date
+
+            worked = holiday.employee_id.get_work_days_count(period_start, period_end, domain=[('holiday_id.holiday_status_id.unpaid', '=', True), ('time_type', '=', 'leave')])
+            left = holiday.employee_id.get_leaves_day_count(period_start, period_end, domain=[('holiday_id.holiday_status_id.unpaid', '=', True), ('time_type', '=', 'leave')])
+            prorata = worked / (left + worked) if worked else 0
+
+            values['number_of_days_temp'] = holiday.number_of_days_temp + leaves * prorata
+
+            if holiday.holiday_status_id.balance_limit > 0:
+                values['number_of_days_temp'] = min(holiday.number_of_days_temp, holiday.holiday_status_id.balance_limit)
+
+            holiday.write(values)
+
     @api.multi
     @api.depends('number_of_days_temp')
     def _compute_number_of_days(self):
         for holiday in self:
-            holiday.number_of_days = holiday.number_of_days_temp
+            if holiday.holiday_status_id.balance_limit > 0:
```

Comment:
`holiday_status_id` should be in the dependencies. But not `holiday_status_id.balance_limit` otherwise changing the limit of a leave type will trigger the computation of all its allocations.